### PR TITLE
[actix-files] Allow user defined guards for NamedFile (actix#1113)

### LINKED ISSUE #1113: NamedFile can only be returned from a GET request
I've got a route that returns a NamedFile, and it works fine with GET requests.  But now I need to update some database state before I return the file, so I need a POST request.  But when I change my route to POST, I get a 405 error from actix_web with a body "This resource only supports GET and HEAD."  
I've traced the execution through my handler function up until I return the opened NamedFile to actix, so I know it's actually getting into my code and we're successfully opening the file handle.
Is it a requirement to only use NamedFile with GET?

### PR DESCRIPTION
**Pull request by naim94a**
I though this could be a good first issue, please tell me if you would have taken another approach implementing this. 
Should fix #1113.

### CODE REVIEW

File: actix-files/src/named.rs [hunk 1]
```diff
@@ -37,7 +38,6 @@ impl Default for Flags {
 }
 
 /// A file with an associated name.
-#[derive(Debug)]
```
Comment (JohnTitor) [inline actix-files/src/named.rs]: I don't think it's a good idea to remove the Debug implementation.

File: actix-files/src/named.rs [hunk 2]
```diff
@@ -216,6 +218,14 @@ impl NamedFile {
         self
     }
 
+    ///Specifies custom guards.
+    /// 
+    ///Default guards allow GET, HEAD.
```
Comment (JohnTitor) [inline actix-files/src/named.rs]:
```suggestion
    /// Default guards allow GET, HEAD. 
```

File: actix-files/src/named.rs [hunk 3]
```diff
@@ -216,6 +218,14 @@ impl NamedFile {
         self
     }
 
+    ///Specifies custom guards.
```
Comment (JohnTitor) [inline actix-files/src/named.rs]:
```suggestion
    /// Specifies custom guards.
```

File: actix-files/src/named.rs [hunk 4]
```diff
@@ -324,14 +334,20 @@ impl Responder for NamedFile {
...
-            Method::HEAD | Method::GET => (),
-            _ => {
-                return Ok(HttpResponse::MethodNotAllowed()
-                    .header(header::CONTENT_TYPE, "text/plain")
-                    .header(header::ALLOW, "GET, HEAD")
-                    .body("This resource only supports GET and HEAD."));
+        let are_guards_satisfied = if let Some(ref guard) = self.guard {
+            (**guard).check(req.head())
+        }
+        else {
```
Comment (JohnTitor) [inline actix-files/src/named.rs]: These lines(L339 and L340) can be one line.

**Review by JohnTitor — COMMENTED** [review]
I'm not sure the current implementation is good, I'd like to hear other's opinions. Also, we should add the test to make sure that works well.

**Comment by naim94a** [comment]
> I'm not sure the current implementation is good, I'd like to hear other's opinions. Also, we should add the test to make sure that works well.

The other approach I thought about was to make NamedFile generic: 
```rust
struct NamedFile<G: Guard>{ ... }
```

**Comment by fafhrd91** [comment]
i think method check should be moved to `Files` type. `NamedFile` should not assume or imply any guard checks.

**Comment by naim94a** [comment]
@fafhrd91 Do you think its a good idea to remove the existing checks from `NamedFile` (checks if the method is GET/HEAD)? Or should the existing checks be bypassed when using `Files`?

**Comment by fafhrd91** [comment]
I think we should move guard like checks to Files. but HEAD handling should stay.

**Review by fafhrd91 — APPROVED** [review]
could you also add entry to changes.md

**Pull request merged**